src/Generating/RandomizeStrategy.cpp

```diff
@@ -0,0 +1,264 @@
+
+// RandomizeStrategy.cpp
+
+// Implements the various classes descending from cPiece::cRandomizeStrategy
+
+#include "Globals.h"
+#include "RandomizeStrategy.h"
+#include "../Noise/Noise.h"
+#include "../Item.h"
+
+
+
+
+
+// Constant that is added to random seed
+static const int SEED_OFFSET = 135 * 13;
+
+
+
+
+
+// Emit a warning if the first param is true
+#define CONDWARNING(ShouldLog, Fmt, ...) \
+	do { \
+		if (ShouldLog) \
+		{ \
+			LOGWARNING(Fmt, __VA_ARGS__); \
+		} \
+	} while (false)
+
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+/** A randomize strategy that pseudo-randomly replaces wheat to various crops of different growth stages. */
+class cRandomizeStrategyCrops:
+	public cPiece::cRandomizeStrategy
+{
+public:
+	cRandomizeStrategyCrops(void) :
+		m_Seed()
+	{
+	}
+
+	virtual bool InitializeFromString(const AString & a_Params, bool a_LogWarnings) override
+	{
+		AString Params = a_Params;
+
+		auto idxPipe = Params.find('|');
+		if (idxPipe != AString::npos)
+		{
+			auto blockToReplace = cItem();
+			AString blockToReplaceStr = Params.substr(0, idxPipe);
+			StringToItem(blockToReplaceStr, blockToReplace);
+			if (blockToReplace.IsEmpty())
+			{
+				m_BlockToReplace = blockToReplace.m_ItemType;
+			}
+			else
+			{
+				// Failed to parse block to replace:
+				CONDWARNING(a_LogWarnings, "Cannot parse block to replace from string \"%s\"!", blockToReplaceStr.c_str());
```

Comment:
I believe a return false is missing here.

Reply:
Block to replace is optional param like I showed in examples, default is wheat in this case of crops. Maybe after I rename it to Basic not Crops it should be required param.

Reply:
But in this case the warning is shown when a block is actually specified but it failed to parse. Currently if the parsing fails `m_BlockToReplace` isn't filled but it still goes ahead right?

Reply:
So if I'd use this string it wouldn't fill `m_BlockToReplace` but it would still go ahead with only a warning:

`["RandomizeStrategy"] = "Crops|someblockthatdoesntexist|beetroots,5;carrots,15;potatoes,16;air,1"`

Reply:
Yes you are right, it was wrong design letting invalid block's be in metadata without warning.